terminology/terminology_processor.bal

```diff
@@ -0,0 +1,1661 @@
+                    return codeSystemArray.slice(offset, offset + count).clone();
+                } else if total >= offset {
+                    return codeSystemArray.slice(offset).clone();
+                } else {
+                    return [];
+                }
+            }
+        }
+    }
+
+    # Search for Value Sets for the provided search parameters.
+    # Allowed search parameters: are name, title, url, version, status and son on.
+    #
+    # + params - List of search parameters, should be passed as map of string arrays  
+    # + return - Return array of ValueSet data if success, return FHIR error if the request contains unsupported search parameters
+    # and for any other processing errors
+    public isolated function searchValueSets(map<r4:RequestSearchParameter[]> params) returns r4:FHIRError|i4:ValueSet[] {
+        lock {
+            map<r4:RequestSearchParameter[]> searchParameters = params.clone();
+            int count = TERMINOLOGY_SEARCH_DEFAULT_COUNT;
+            if searchParameters.hasKey("_count") {
+                int|error y = langint:fromString(searchParameters.get("_count")[0].value);
+                if y is int {
+                    count = y;
+                }
+                _ = searchParameters.remove("_count");
+            }
+
+            int offset = 0;
+            if searchParameters.hasKey("_offset") {
+                int|error y = langint:fromString(searchParameters.get("_offset")[0].value);
+                if y is int {
+                    offset = y;
+                }
+                _ = searchParameters.remove("_offset");
+            }
+
+            if count > TERMINOLOGY_SEARCH_MAXIMUM_COUNT {
+                return r4:createFHIRError(
+                string `Requested size of the response: ${count.toBalString()} is too large`,
+                r4:ERROR,
+                r4:PROCESSING_NOT_SUPPORTED,
+                diagnostic = string `Allowed maximum size of output is: ${TERMINOLOGY_SEARCH_MAXIMUM_COUNT}; therefore, reduce the value of size parameter accordingly`,
+                errorType = r4:PROCESSING_ERROR,
+                httpStatusCode = http:STATUS_PAYLOAD_TOO_LARGE
+                );
+            }
+
+            // Validate the requested search parameters in the allowed list
+            foreach var param in searchParameters.clone().keys() {
+                if !VALUESETS_SEARCH_PARAMS.hasKey(param) {
+                    return r4:createFHIRError(
+                string `This search parameter is not implemented yet: ${param}`,
+                r4:ERROR,
+                r4:PROCESSING_NOT_SUPPORTED,
+                diagnostic = string `Allowed search parameters: ${VALUESETS_SEARCH_PARAMS.keys().toBalString()}`,
+                errorType = r4:VALIDATION_ERROR);
+                }
+            }
+
+            if self.finder is Finder {
+                return (<Finder>self.finder).searchValueSet(searchParameters.clone(), offset = offset, count = count);
```

Comment:
no need to clone the searchParameters again?

Reply:
If we remove that, getting the below error
```
invalid attempt to transfer out a value from a 'lock' statement with restricted variable usage: expected an isolated expression(BCE3959)
```